File: bindings/takumi-render-uniffi/csharp/Takumi.Render.UniFFI/Renderer.cs
using Generated = Takumi.Render.UniFFI.Generated;

namespace Takumi.Render.UniFFI;

public sealed class Renderer : IDisposable
{
    private readonly Generated.Renderer _inner;

    public Renderer()
    {
        _inner = new Generated.Renderer();
    }

    public void AddFontBytes(byte[] bytes)
    {
        Invoke(() => _inner.AddFontBytes(bytes));
    }

    public void AddFontDirectory(string path)
    {
        Invoke(() => _inner.AddFontDirectory(path));
    }

    public void AddFontFile(string path)
    {
        Invoke(() => _inner.AddFontFile(path));
    }

    public void AddSearchPath(string path)
    {
        Invoke(() => _inner.AddSearchPath(path));
    }

    public void ClearCaches()
    {
        Invoke(_inner.ClearCaches);
    }

    public MeasuredLayout Measure(RenderRequest request)
    {
        return MeasuredLayout.FromGenerated(Invoke(() => _inner.Measure(request.ToGenerated())));
    }

    public RenderedImage Render(RenderRequest request)
    {
        return RenderedImage.FromGenerated(Invoke(() => _inner.Render(request.ToGenerated())));
    }

    public RenderedImage RenderToFile(RenderRequest request, string outputPath)
    {
        return RenderedImage.FromGenerated(
            Invoke(() => _inner.RenderToFile(request.ToGenerated(), outputPath))
        );
    }

    public void Dispose()
    {
        _inner.Dispose();
    }

    private static void Invoke(Action action)
    {
        try
        {
            action();
        }
        catch (Exception exception)
            when (exception is Generated.RendererException or RendererException)
        {
            throw TranslateException(exception);
        }
    }

    private static T Invoke<T>(Func<T> action)
    {
        try
        {
            return action();
        }
        catch (Exception exception)
            when (exception is Generated.RendererException or RendererException)
        {
            throw TranslateException(exception);
        }
    }

    private static RendererException TranslateException(Exception exception) =>
        exception switch
        {
            RendererException rendererException => rendererException,
            Generated.RendererException generated => RendererException.FromGenerated(generated),
            _ => new RendererException(RendererErrorKind.Unknown, exception.Message, exception),
        };
}

File: bindings/takumi-render-uniffi/jvm/src/main/kotlin/io/github/zhjk/takumi/render/uniffi/Renderer.kt
package io.github.zhjk.takumi.render.uniffi

import io.github.zhjk.takumi.render.uniffi.generated.Renderer as GeneratedRenderer
import io.github.zhjk.takumi.render.uniffi.generated.TakumiRenderUniffiNativeLoader

class Renderer : AutoCloseable {
    private val inner: GeneratedRenderer

    init {
        TakumiRenderUniffiNativeLoader.ensureLoaded()
        inner = GeneratedRenderer()
    }

    fun addFontBytes(bytes: ByteArray) {
        invokeVoid { inner.addFontBytes(bytes) }
    }

    fun addFontDirectory(path: String) {
        invokeVoid { inner.addFontDirectory(path) }
    }

    fun addFontFile(path: String) {
        invokeVoid { inner.addFontFile(path) }
    }

    fun addSearchPath(path: String) {
        invokeVoid { inner.addSearchPath(path) }
    }

    fun clearCaches() {
        invokeVoid { inner.clearCaches() }
    }

    fun measure(request: RenderRequest): MeasuredLayout =
        MeasuredLayout.fromGenerated(invoke { inner.measure(request.toGenerated()) })

    fun render(request: RenderRequest): RenderedImage =
        RenderedImage.fromGenerated(invoke { inner.render(request.toGenerated()) })

    fun renderToFile(request: RenderRequest, outputPath: String): RenderedImage =
        RenderedImage.fromGenerated(invoke { inner.renderToFile(request.toGenerated(), outputPath) })

    override fun close() {
        inner.close()
    }

    private inline fun invokeVoid(action: () -> Unit) {
        try {
            action()
        } catch (throwable: Throwable) {
            throw translateThrowable(throwable)
        }
    }

    private inline fun <T> invoke(action: () -> T): T =
        try {
            action()
        } catch (throwable: Throwable) {
            throw translateThrowable(throwable)
        }

    private fun translateThrowable(throwable: Throwable): RendererException = when (throwable) {
        is RendererException -> throwable
        else -> RendererException.fromGenerated(throwable)
    }
}
